pennylane/labs/resource_estimation/templates/subroutines.py

```diff
@@ -477,6 +477,408 @@ def default_controlled_resource_decomp(
+            self.wires = Wires(wires)
+            self.num_wires = len(self.wires)
+        else:
+            self.wires = base_op.wires or None
+            self.num_wires = base_op.num_wires
+
+        super().__init__(wires=wires)
+
+    @property
+    def resource_params(self):
+        r"""Returns a dictionary containing the minimal information needed to compute the resources.
+
+        Returns:
+            dict: A dictionary containing the resource parameters:
+                * base_cmpr_op (CompressedResourceOp): A compressed resource operator, corresponding
+                  to the operator that we will be applying controlled powers of.
+                * num_iter (int): the number of mid-circuit measurements made to read out the phase
+        """
+        return {"base_cmpr_op": self.base_cmpr_op, "num_iter": self.num_iter}
+
+    @classmethod
+    def resource_rep(cls, base_cmpr_op, num_iter) -> CompressedResourceOp:
+        r"""Returns a compressed representation containing only the parameters of
+        the Operator that are needed to compute the resources.
+
+        Args:
+            base_cmpr_op (CompressedResourceOp): A compressed resource operator, corresponding
+                to the operator that we will be applying controlled powers of.
+            num_iter (int): the number of mid-circuit measurements made to read out the phase
+
+        Returns:
+            CompressedResourceOp: the operator in a compressed representation
+        """
+        return CompressedResourceOp(cls, {"base_cmpr_op": base_cmpr_op, "num_iter": num_iter})
+
+    @classmethod
+    def default_resource_decomp(cls, base_cmpr_op, num_iter, **kwargs):
+        r"""Returns a list representing the resources of the operator. Each object in the list represents a gate and the
+        number of times it occurs in the circuit.
+
+        Args:
+            base_cmpr_op (CompressedResourceOp): A compressed resource operator, corresponding
+                to the operator that we will be applying controlled powers of.
+            num_iter (int): the number of mid-circuit measurements made to read out the phase
+
+        Resources:
+            The resources are obtained following the constuction from `arXiv:quant-ph/0610214v3
+            <https://arxiv.org/pdf/quant-ph/0610214v3>`_.
+
+        Returns:
+            list[GateCount]: A list of GateCount objects, where each object
+            represents a specific quantum gate and the number of times it appears
+            in the decomposition.
+        """
+        gate_counts = [
+            GateCount(resource_rep(re.ResourceHadamard), 2 * num_iter),
+            AllocWires(1),
+        ]
+
+        # Here we want to use this particular decomposition, not any random one the user might override
+        gate_counts += ResourceControlledSequence.default_resource_decomp(base_cmpr_op, num_iter)
+
+        num_phase_gates = num_iter * (num_iter - 1) // 2
+        gate_counts.append(
```

Comment:
Is this from a specific equation in the paper? If yes, will be great to cite that.